jetty-server/src/main/java/org/eclipse/jetty/server/PlainOrSslConnectionFactory.java

```diff
@@ -0,0 +1,192 @@
+
+/**
+ * <p>A ConnectionFactory whose connections detect whether the first bytes are
+ * TLS bytes and upgrades to either a TLS connection or to a plain connection.</p>
+ */
+public class PlainOrSslConnectionFactory extends AbstractConnectionFactory
+{
+    private static final Logger LOG = Log.getLogger(PlainOrSslConnection.class);
+    private static final int TLS_ALERT_FRAME_TYPE = 0x15;
+    private static final int TLS_HANDSHAKE_FRAME_TYPE = 0x16;
+    private static final int TLS_MAJOR_VERSION = 3;
+
+    private final SslConnectionFactory sslConnectionFactory;
+    private final String plainProtocol;
+
+    /**
+     * <p>Creates a new plain or TLS ConnectionFactory.</p>
+     * <p>If {@code plainProtocol} is {@code null}, and the first bytes are not TLS, then
+     * {@link #unknownProtocol(ByteBuffer, EndPoint)} is called; applications may override its
+     * behavior (by default it closes the EndPoint) for example by writing a minimal response. </p>
+     *
+     * @param sslConnectionFactory The SslConnectionFactory to use if the first bytes are TLS
+     * @param plainProtocol        the protocol of the ConnectionFactory to use if the first bytes are not TLS, or null.
+     */
+    public PlainOrSslConnectionFactory(SslConnectionFactory sslConnectionFactory, String plainProtocol)
+    {
+        super("plain|ssl");
+        this.sslConnectionFactory = sslConnectionFactory;
+        this.plainProtocol = plainProtocol;
+    }
+
+    @Override
+    public Connection newConnection(Connector connector, EndPoint endPoint)
+    {
+        return configure(new PlainOrSslConnection(endPoint, connector), connector, endPoint);
+    }
+
+    /**
+     * @param buffer The buffer with the first bytes of the connection
+     * @return whether the bytes seem TLS bytes
+     */
+    protected boolean seemsTLS(ByteBuffer buffer)
+    {
+        int tlsFrameType = buffer.get(0) & 0xFF;
+        int tlsMajorVersion = buffer.get(1) & 0xFF;
+        return (tlsFrameType == TLS_HANDSHAKE_FRAME_TYPE || tlsFrameType == TLS_ALERT_FRAME_TYPE) && tlsMajorVersion == TLS_MAJOR_VERSION;
+    }
+
+    /**
+     * <p>Callback method invoked when {@code plainProtocol} is {@code null}
+     * and the first bytes are not TLS.</p>
+     * <p>This typically happens when a client is trying to connect to a TLS
+     * port using the {@code http} scheme (and not the {@code https} scheme).</p>
+     * <p>This method may be overridden to write back a minimal response such as:</p>
+     * <pre>
+     * HTTP/1.1 400 Bad Request
+     * Content-Length: 35
+     * Content-Type: text/plain; charset=UTF8
+     * Connection: close
+     *
+     * Plain HTTP request sent to TLS port
+     * </pre>
+     *
+     * @param buffer   The buffer with the first bytes of the connection
+     * @param endPoint The connection EndPoint object
+     * @see #seemsTLS(ByteBuffer)
+     */
+    protected void unknownProtocol(ByteBuffer buffer, EndPoint endPoint)
+    {
+        endPoint.close();
```

Comment:
Is it possible to write an implementation that does a 302 redirect to the https version?
Shouldn't our implementation do the 400 response by default? (instead of just closing the endpoint)?
Shouldn't all implementations of this close the endpoint afterwords?

Reply:
Yes, it is possible to write a HTTP response, see test case that writes a 400.

Responding 400 by default implies that the client is using http/1.1, which may not be the case (e.g. fcgi or h2c).

I don't understand the 3rd point. Overrides of `unknownProtocol()` typically close the connection, but in principle they can also upgrade to a default protocol, etc.

Reply:
A generic 302 redirect is easy enough.
But an intelligent one would require knowledge from the HTTP request line, right?

Reply:
@gregw some examples (note that these are curl requests to port 443, but scheme `http`)

nginx behavior:

```
[joakim@hyperion tmp]$ curl -D - http://wikipedia.org:443/
HTTP/1.1 400 Bad Request
Server: nginx/1.13.6
Date: Thu, 31 Jan 2019 22:08:45 GMT
Content-Type: text/html
Content-Length: 271
Connection: close

<html>
<head><title>400 The plain HTTP request was sent to HTTPS port</title></head>
<body bgcolor="white">
<center><h1>400 Bad Request</h1></center>
<center>The plain HTTP request was sent to HTTPS port</center>
<hr><center>nginx/1.13.6</center>
</body>
</html>
```

and

```
[joakim@hyperion tmp]$ curl -D - http://eclipse.org:443/
HTTP/1.1 400 Bad Request
Server: nginx
Date: Thu, 31 Jan 2019 22:09:30 GMT
Content-Type: text/html
Content-Length: 264
Connection: close

<html>
<head><title>400 The plain HTTP request was sent to HTTPS port</title></head>
<body bgcolor="white">
<center><h1>400 Bad Request</h1></center>
<center>The plain HTTP request was sent to HTTPS port</center>
<hr><center>nginx</center>
</body>
</html>
```

And Apache behavior ...

```
[joakim@hyperion tmp]$ curl -D - http://ietf.org:443/
HTTP/1.1 400 Bad Request
Date: Thu, 31 Jan 2019 22:08:55 GMT
Server: Apache
Vary: accept-language,accept-charset,Accept-Encoding
Accept-Ranges: bytes
Strict-Transport-Security: max-age=31536000
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Connection: close
Content-Type: text/html; charset=utf-8
Content-Language: en
Expires: Thu, 31 Jan 2019 22:08:55 GMT

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<title>Bad request!</title>
<link rev="made" href="mailto:ietf-action@ietf.org" />
<style type="text/css"><!--/*--><![CDATA[/*><!--*/ 
    body { color: #000000; background-color: #FFFFFF; }
    a:link { color: #0000CC; }
    p, address {margin-left: 3em;}
    span {font-size: smaller;}
/*]]>*/--></style>
</head>

<body>
<h1>Bad request!</h1>
<p>


    Your browser (or proxy) sent a request that
    this server could not understand.

</p>
<p>
If you think this is a server error, please contact
the <a href="mailto:ietf-action@ietf.org">webmaster</a>.

</p>

<h2>Error 400</h2>
<address>
  <a href="/">www7.ietf.org</a><br />
  <span>Apache</span>
</address>
</body>
</html>
```

Reply:
Perhaps a configurable field?
`reportUnknownProtocol=true` would produce a default HTML error page, customizable by overriding some method, while `false` would just close the connection.

Reply:
If you want to avoid another factory, then we should read at least 3 bytes, check for ssl (perhaps only in first 2 bytes) and if not found, get for `GET` to decide if we send a 400 or just close.
I guess the problem then is with any other protocol that has an opening handshake of only 1 or 2 bytes! So perhaps the number of bytes to check should be configurable with 3 being the default.

In short, I'm very reluctant to send HTTP responses unless we have checked it is likely to be HTTP.

Reply:
@gregw perhaps the method name is misleading. It is not unknown, the only certain thing is that it's not the TLS protocol. There is no need for a second `ConnectionFactory` because there is already the `plainProtocol` name and we can retrieve the `ConnectionFactory` from the name.
It would be easy to check if it's HTTP by parsing the `buffer`.
What you're suggesting is to ensure that the `buffer` has at least 3 bytes before upgrading or calling `unknownProtocol`?

Reply:
Those replies are sent, regardless of client.

Examples:

(A tab character only)

```
$ echo -e "\t" | netcat w3c.org 443
HTTP/1.1 400 Bad Request
Server: nginx/1.10.3
Date: Fri, 01 Feb 2019 23:43:53 GMT
Content-Type: text/html
Content-Length: 173
Connection: close

<html>
<head><title>400 Bad Request</title></head>
<body bgcolor="white">
<center><h1>400 Bad Request</h1></center>
<hr><center>nginx/1.10.3</center>
</body>
</html>
```

(A single byte of 0x00)

```
$ echo -e '\x00' | netcat wikipedia.org 443
HTTP/1.1 400 Bad Request
Server: nginx/1.13.6
Date: Fri, 01 Feb 2019 23:46:14 GMT
Content-Type: text/html
Content-Length: 173
Connection: close

<html>
<head><title>400 Bad Request</title></head>
<body bgcolor="white">
<center><h1>400 Bad Request</h1></center>
<hr><center>nginx/1.13.6</center>
</body>
</html>
```

Reply:
@joakime those are not the same 400's   Specifically they don't say "The plain HTTP request was sent to HTTPS port".
I still think that if we are sniffing characters, then checking for a `G`, `E` and `T`  is a reasonable thing to do before concluding that "The plain HTTP request was sent to HTTPS port".     If we don't see a HTTP request, then I'm fine with just closing the connection, but can live with a blank 400 as well.

Reply:
@sbordet, I'm suggesting that we check at least 2, maybe 3 characters and then:
 * If they are TLS, then we do TLS.   
 * If they are not TLS and we have configured another protocol then the do the other protocol.
 * If they are not TLS and we have no other protocol and they are `G` & `E` then we send a 400 with a message saying "HTTP request sent to HTTPS port"
 * If they are not TLS and we have no other protocol and they are not `G` & `E` then we either just close or send an empty 400 and then close.

For the name, how about `optionalTLS`? 

(Note I'm never sure if we should be saying TLS or SSL... so if we want to stay with SSL for consistency, then s/TLS/SSL/g above)